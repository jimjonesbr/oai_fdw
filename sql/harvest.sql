CREATE FOREIGN TABLE IF NOT EXISTS ulb_oai_dc (
  id text                OPTIONS (oai_node 'identifier'), 
  xmldoc xml             OPTIONS (oai_node 'content'), 
  sets text[]            OPTIONS (oai_node 'setspec'), 
  updatedate timestamp   OPTIONS (oai_node 'datestamp'),
  format text            OPTIONS (oai_node 'metadataprefix')
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');

/* Target table without schema */
CALL OAI_HarvestTable('ulb_oai_dc','clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);
SELECT count(*) FROM clone_ulb_oai_dc;

-- Existing table will be upserted 
CALL OAI_HarvestTable('ulb_oai_dc','clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00');
SELECT count(*) FROM clone_ulb_oai_dc;

/* Target table with specific schema */
CALL OAI_HarvestTable('ulb_oai_dc','oai_schema.clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);
SELECT count(*) FROM oai_schema.clone_ulb_oai_dc;

-- Existing table will be upserted 
CALL OAI_HarvestTable('ulb_oai_dc','oai_schema.clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00');
SELECT count(*) FROM oai_schema.clone_ulb_oai_dc;

/* Target table containing special characters */
CALL OAI_HarvestTable('ulb_oai_dc','oai_schema."clone-ulb:oai/dc"', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);
SELECT count(*) FROM oai_schema."clone-ulb:oai/dc";

/* Target table containing special characters without schema */
CALL OAI_HarvestTable('ulb_oai_dc','"clone-ulb:oai/dc"', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);
SELECT count(*) FROM "clone-ulb:oai/dc";

/* WARNING: Foreign table without identifier */
CREATE FOREIGN TABLE IF NOT EXISTS table_without_identifier (
  xmldoc xml             OPTIONS (oai_node 'content'), 
  sets text[]            OPTIONS (oai_node 'setspec'), 
  updatedate timestamp   OPTIONS (oai_node 'datestamp'),
  format text            OPTIONS (oai_node 'metadataprefix')
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');
 
CALL OAI_HarvestTable('table_without_identifier','clone_table_without_identifier', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true); 

/* EXCEPTION: oai_table does not exist */
CALL OAI_HarvestTable('foo','clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);

/* EXCEPTION: end date smaller than start date */
CALL OAI_HarvestTable('ulb_oai_dc','clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-09-01 00:00:00',true,true);

/* EXCEPTION: Foreign table without datestamp */
CREATE FOREIGN TABLE IF NOT EXISTS table_without_datestamp (
  id text                OPTIONS (oai_node 'identifier'), 
  xmldoc xml             OPTIONS (oai_node 'content'), 
  sets text[]            OPTIONS (oai_node 'setspec'), 
  format text            OPTIONS (oai_node 'metadataprefix')
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');

CALL OAI_HarvestTable('table_without_datestamp', 'clone_table_without_datestamp', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true); 

/* EXCEPTION: Foreign table without datestamp and identifier*/
CREATE FOREIGN TABLE IF NOT EXISTS table_without_datestamp_identifier (
  xmldoc xml             OPTIONS (oai_node 'content'), 
  sets text[]            OPTIONS (oai_node 'setspec'), 
  format text            OPTIONS (oai_node 'metadataprefix')
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');

CALL OAI_HarvestTable('table_without_datestamp_identifier', 'clone_table_without_datestamp_identifier', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true); 

/* EXCEPTION: Foreign table without any oai_node */
CREATE FOREIGN TABLE IF NOT EXISTS table_without_oai_node (
  xmldoc xml, 
  sets text[], 
  format text
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');

CALL OAI_HarvestTable('table_without_oai_node', 'clone_table_without_oai_node', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true); 
