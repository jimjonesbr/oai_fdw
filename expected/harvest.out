CREATE FOREIGN TABLE IF NOT EXISTS ulb_oai_dc (
  id text                OPTIONS (oai_node 'identifier'), 
  xmldoc xml             OPTIONS (oai_node 'content'), 
  sets text[]            OPTIONS (oai_node 'setspec'), 
  updatedate timestamp   OPTIONS (oai_node 'datestamp'),
  format text            OPTIONS (oai_node 'metadataprefix')
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');
/* Target table without schema */
CALL OAI_HarvestTable('ulb_oai_dc','clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);
INFO:  target table "public.clone_ulb_oai_dc" created
INFO:  1 records from "public.ulb_oai_dc" updated or inserted into "public.clone_ulb_oai_dc" [2022-10-01 00:00:00 - 2022-10-06 00:00:00]
INFO:  57 records from "public.ulb_oai_dc" updated or inserted into "public.clone_ulb_oai_dc" [2022-10-06 00:00:00 - 2022-10-11 00:00:00]
INFO:  32 records from "public.ulb_oai_dc" updated or inserted into "public.clone_ulb_oai_dc" [2022-10-11 00:00:00 - 2022-10-16 00:00:00]
INFO:  OAI harvester complete ("public.ulb_oai_dc" -> "public.clone_ulb_oai_dc"): 90 records affected [2022-10-01 00:00:00 - 2022-10-17 00:00:00]
SELECT count(*) FROM clone_ulb_oai_dc;
 count 
-------
    90
(1 row)

-- Existing table will be upserted 
CALL OAI_HarvestTable('ulb_oai_dc','clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00');
INFO:  OAI harvester complete ("public.ulb_oai_dc" -> "public.clone_ulb_oai_dc"): 90 records affected [2022-10-01 00:00:00 - 2022-10-17 00:00:00]
SELECT count(*) FROM clone_ulb_oai_dc;
 count 
-------
    90
(1 row)

/* Target table with specific schema */
CALL OAI_HarvestTable('ulb_oai_dc','oai_schema.clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);
ERROR:  schema "oai_schema" does not exist
CONTEXT:  SQL statement "CREATE TABLE oai_schema.clone_ulb_oai_dc AS SELECT id, xmldoc, sets, updatedate, format FROM public.ulb_oai_dc WITH NO DATA;"
PL/pgSQL function oai_harvesttable(text,text,interval,timestamp without time zone,timestamp without time zone,boolean,boolean) line 72 at EXECUTE
SELECT count(*) FROM oai_schema.clone_ulb_oai_dc;
ERROR:  relation "oai_schema.clone_ulb_oai_dc" does not exist
LINE 1: SELECT count(*) FROM oai_schema.clone_ulb_oai_dc;
                             ^
-- Existing table will be upserted 
CALL OAI_HarvestTable('ulb_oai_dc','oai_schema.clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00');
ERROR:  schema "oai_schema" does not exist
CONTEXT:  SQL statement "CREATE TABLE oai_schema.clone_ulb_oai_dc AS SELECT id, xmldoc, sets, updatedate, format FROM public.ulb_oai_dc WITH NO DATA;"
PL/pgSQL function oai_harvesttable(text,text,interval,timestamp without time zone,timestamp without time zone,boolean,boolean) line 72 at EXECUTE
SELECT count(*) FROM oai_schema.clone_ulb_oai_dc;
ERROR:  relation "oai_schema.clone_ulb_oai_dc" does not exist
LINE 1: SELECT count(*) FROM oai_schema.clone_ulb_oai_dc;
                             ^
/* Target table containing special characters */
CALL OAI_HarvestTable('ulb_oai_dc','oai_schema."clone-ulb:oai/dc"', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);
ERROR:  schema "oai_schema" does not exist
CONTEXT:  SQL statement "CREATE TABLE oai_schema."clone-ulb:oai/dc" AS SELECT id, xmldoc, sets, updatedate, format FROM public.ulb_oai_dc WITH NO DATA;"
PL/pgSQL function oai_harvesttable(text,text,interval,timestamp without time zone,timestamp without time zone,boolean,boolean) line 72 at EXECUTE
SELECT count(*) FROM oai_schema."clone-ulb:oai/dc";
ERROR:  relation "oai_schema.clone-ulb:oai/dc" does not exist
LINE 1: SELECT count(*) FROM oai_schema."clone-ulb:oai/dc";
                             ^
/* Target table containing special characters without schema */
CALL OAI_HarvestTable('ulb_oai_dc','"clone-ulb:oai/dc"', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);
INFO:  target table "public."clone-ulb:oai/dc"" created
INFO:  1 records from "public.ulb_oai_dc" updated or inserted into "public."clone-ulb:oai/dc"" [2022-10-01 00:00:00 - 2022-10-06 00:00:00]
INFO:  57 records from "public.ulb_oai_dc" updated or inserted into "public."clone-ulb:oai/dc"" [2022-10-06 00:00:00 - 2022-10-11 00:00:00]
INFO:  32 records from "public.ulb_oai_dc" updated or inserted into "public."clone-ulb:oai/dc"" [2022-10-11 00:00:00 - 2022-10-16 00:00:00]
INFO:  OAI harvester complete ("public.ulb_oai_dc" -> "public."clone-ulb:oai/dc""): 90 records affected [2022-10-01 00:00:00 - 2022-10-17 00:00:00]
SELECT count(*) FROM "clone-ulb:oai/dc";
 count 
-------
    90
(1 row)

/* WARNING: Foreign table without identifier */
CREATE FOREIGN TABLE IF NOT EXISTS table_without_identifier (
  xmldoc xml             OPTIONS (oai_node 'content'), 
  sets text[]            OPTIONS (oai_node 'setspec'), 
  updatedate timestamp   OPTIONS (oai_node 'datestamp'),
  format text            OPTIONS (oai_node 'metadataprefix')
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');
 
CALL OAI_HarvestTable('table_without_identifier','clone_table_without_identifier', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true); 
WARNING:  foreign table "public.table_without_identifier" has no identifier column. It is strongly recommended to map the OAI identifier to a column, as it can ensure that records are not duplicated
INFO:  target table "public.clone_table_without_identifier" created
INFO:  1 records from "public.table_without_identifier" updated or inserted into "public.clone_table_without_identifier" [2022-10-01 00:00:00 - 2022-10-06 00:00:00]
INFO:  57 records from "public.table_without_identifier" updated or inserted into "public.clone_table_without_identifier" [2022-10-06 00:00:00 - 2022-10-11 00:00:00]
INFO:  32 records from "public.table_without_identifier" updated or inserted into "public.clone_table_without_identifier" [2022-10-11 00:00:00 - 2022-10-16 00:00:00]
INFO:  OAI harvester complete ("public.table_without_identifier" -> "public.clone_table_without_identifier"): 90 records affected [2022-10-01 00:00:00 - 2022-10-17 00:00:00]
/* EXCEPTION: oai_table does not exist */
CALL OAI_HarvestTable('foo','clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true);
ERROR:  foreign table "public.foo" does not exist
CONTEXT:  PL/pgSQL function oai_harvesttable(text,text,interval,timestamp without time zone,timestamp without time zone,boolean,boolean) line 57 at RAISE
/* EXCEPTION: end date smaller than start date */
CALL OAI_HarvestTable('ulb_oai_dc','clone_ulb_oai_dc', interval '5 days', '2022-10-01 00:00:00', '2022-09-01 00:00:00',true,true);
ERROR:  invalid time window. The end date [Sat Oct 01 00:00:00 2022] lies before the start date [Thu Sep 01 00:00:00 2022]
CONTEXT:  PL/pgSQL function oai_harvesttable(text,text,interval,timestamp without time zone,timestamp without time zone,boolean,boolean) line 26 at RAISE
/* EXCEPTION: Foreign table without datestamp */
CREATE FOREIGN TABLE IF NOT EXISTS table_without_datestamp (
  id text                OPTIONS (oai_node 'identifier'), 
  xmldoc xml             OPTIONS (oai_node 'content'), 
  sets text[]            OPTIONS (oai_node 'setspec'), 
  format text            OPTIONS (oai_node 'metadataprefix')
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');
CALL OAI_HarvestTable('table_without_datestamp', 'clone_table_without_datestamp', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true); 
ERROR:  foreign table "public.table_without_datestamp" has no datestamp column
CONTEXT:  PL/pgSQL function oai_harvesttable(text,text,interval,timestamp without time zone,timestamp without time zone,boolean,boolean) line 65 at RAISE
/* EXCEPTION: Foreign table without datestamp and identifier*/
CREATE FOREIGN TABLE IF NOT EXISTS table_without_datestamp_identifier (
  xmldoc xml             OPTIONS (oai_node 'content'), 
  sets text[]            OPTIONS (oai_node 'setspec'), 
  format text            OPTIONS (oai_node 'metadataprefix')
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');
CALL OAI_HarvestTable('table_without_datestamp_identifier', 'clone_table_without_datestamp_identifier', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true); 
ERROR:  foreign table "public.table_without_datestamp_identifier" has no datestamp column
CONTEXT:  PL/pgSQL function oai_harvesttable(text,text,interval,timestamp without time zone,timestamp without time zone,boolean,boolean) line 65 at RAISE
/* EXCEPTION: Foreign table without any oai_node */
CREATE FOREIGN TABLE IF NOT EXISTS table_without_oai_node (
  xmldoc xml, 
  sets text[], 
  format text
 ) SERVER oai_server_ulb OPTIONS (metadataprefix 'oai_dc');
CALL OAI_HarvestTable('table_without_oai_node', 'clone_table_without_oai_node', interval '5 days', '2022-10-01 00:00:00', '2022-10-17 00:00:00',true,true); 
ERROR:  foreign table "public.table_without_oai_node" has no datestamp column
CONTEXT:  PL/pgSQL function oai_harvesttable(text,text,interval,timestamp without time zone,timestamp without time zone,boolean,boolean) line 65 at RAISE
